internal/buildpackage/builder_test.go

```diff
@@ -464,4 +491,108 @@ func testPackageBuilder(t *testing.T, when spec.G, it spec.S) {
 			buildpackExists("bp.1.id", "bp.1.version")
 		})
 	})
+
+	when("#SaveAsFile", func() {
+		it("sets metadata", func() {
+			buildpack1, err := ifakes.NewFakeBuildpack(dist.BuildpackDescriptor{
+				API: api.MustParse("0.2"),
+				Info: dist.BuildpackInfo{
+					ID:      "bp.1.id",
+					Version: "bp.1.version",
+				},
+				Stacks: []dist.Stack{
+					{ID: "stack.id.1"},
+					{ID: "stack.id.2"},
+				},
+				Order: nil,
+			}, 0644)
+			h.AssertNil(t, err)
+
+			subject.SetBuildpack(buildpack1)
+
+			tempFile, err := ioutil.TempFile(tmpDir, "package-*.cnb")
+			h.AssertNil(t, err)
+			h.AssertNil(t, subject.SaveAsFile(tempFile.Name()))
+
+			// config: application/vnd.docker.container.image.v1+json
+			h.AssertOnTarEntry(t, tempFile.Name(),
+				"/blobs/sha256/12664caac120282c7b806ab1c269b11f050a3f635e844c18ea83240a82628eb4",
```

Comment:
I was able to solve this for the buildpack layer and precompute it's SHA. I liked the outcome because it now explicitly expects the compression level (must be set to default) so that it's inline with the docker client.

As for the other SHAs such as manifest and config, I'm not entirely sure how to solve for it without some more complex traversing. I'll keep thinking about it...

Reply:
Updated to traverse `index.json` -> `manifest (json)` -> `config (json)`